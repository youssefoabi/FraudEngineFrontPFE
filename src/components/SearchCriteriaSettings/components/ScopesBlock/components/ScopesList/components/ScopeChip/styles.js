import { apricot, doveGray, eucalyptus, shamrock, silver, trinidad, white } from '@cdiscount/ui-colors';

export default function style() {
  return {
    baseStyle: {
      borderRadius: 5,
      color: white,
      marginLeft: 5,
      marginTop: 10,
      height: 25,
    },
    statusEnableChip: {
      backgroundColor: apricot,
      '&:hover': {
        backgroundColor: apricot,
      },
    },
    statusEnableCurrentChip: {
      backgroundColor: trinidad,
      '&:hover': {
        backgroundColor: trinidad,
      },
    },
    statusDisableChip: {
      backgroundColor: silver,
      '&:hover': {
        backgroundColor: silver,
      },
    },
    statusDisableCurrentChip: {
      backgroundColor: doveGray,
      '&:hover': {
        backgroundColor: doveGray,
      },
    },
    statusValidChip: {
      backgroundColor: shamrock,
      '&:hover': {
        backgroundColor: shamrock,
      },
    },
    statusValidCurrentChip: {
      backgroundColor: eucalyptus,
      '&:hover': {
        backgroundColor: eucalyptus,
      },
    },
  };
}
