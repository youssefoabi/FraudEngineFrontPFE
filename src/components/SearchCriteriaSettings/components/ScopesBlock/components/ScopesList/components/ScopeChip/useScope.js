import { dissoc, findIndex, flatten, isNil, map, pathOr, pipe, prop, values } from 'ramda';
import { isNilOrEmpty } from 'ramda-adjunct';
import { useEffect, useMemo, useState } from 'react';

import clientApi from '../../../../../../../../client-api';

const emptyObject = {};

const useScope = ({ id, handleChange, engine }) => {
  const { scopes: engineScopes, currentScope } = engine || {};
  const [scope, setScope] = useState({});
  const scopeEngine = pathOr({}, [id], engineScopes);
  const idCurrentScope = pathOr({}, ['scopeId'], currentScope);

  useEffect(() => {
    if (isNil(id)) return;

    clientApi()
      .getScope(id)
      .then(prop('data'))
      .then(scopeData => {
        const parameters = pathOr({}, ['parameters'], scopeData);
        setScope({ parameters, ...scopeEngine });
      });
  }, [id, scopeEngine]);

  const onDelete = index => {
    handleChange({ id: 'scopes', value: dissoc(index, engineScopes) });
    if (index === idCurrentScope) handleChange({ id: 'currentScope', value: null });
  };
  const onClick = clickedScope => handleChange({ id: 'currentScope', value: clickedScope });

  const { inputFields } = engine || emptyObject;

  const hasMissingParameters = useMemo(() => {
    if (isNilOrEmpty(inputFields) || isNilOrEmpty(scope)) return true;

    const { parameters: scopeParameters, scopeId } = scope;
    if (isNilOrEmpty(scopeParameters)) return false;

    const engineParameters = pipe(values, map(prop('parameters')), flatten)(inputFields);

    if (isNilOrEmpty(engineParameters)) return true;

    const firstDisabledPrameter = findIndex(scopeParam => {
      if (isNilOrEmpty(scopeParam)) return false;

      const { key } = scopeParam;

      return (
        findIndex(
          enginePram => key === prop('key', enginePram) && scopeId === prop('scopeId', enginePram),
          engineParameters,
        ) === -1
      );
    })(scopeParameters);

    return firstDisabledPrameter !== -1;
  }, [scope, inputFields]);

  const isEnabled = prop('isEnable', scopeEngine);

  return {
    onDelete,
    onClick,
    scope,
    isEnabled,
    hasMissingParameters,
    idCurrentScope,
  };
};

export default useScope;
