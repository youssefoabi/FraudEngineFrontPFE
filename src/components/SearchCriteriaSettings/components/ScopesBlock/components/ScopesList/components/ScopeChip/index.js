import { func, number, shape } from 'prop-types';
import { isNilOrEmpty } from 'ramda-adjunct';
import { prop } from 'ramda';
import { useStyles } from '@cdiscount/backoffice-ui';
import clsx from 'clsx';
import MuiChip from '@material-ui/core/Chip';
import React, { useState } from 'react';

import DeleteScopeConfirmationModal from '../DeleteScopeConfirmationModal';
import styles from './styles';
import useScope from './useScope';

const ScopeChip = ({ id, handleChange, engine }) => {
  const { hasMissingParameters, onDelete, onClick, scope, isEnabled, idCurrentScope } = useScope({
    handleChange,
    engine,
    id,
  });
  const [deletionModalOpen, setDeletionModalOpen] = useState(false);
  const isCurrentScope = id === idCurrentScope;

  const toggleDeletionModal = () => {
    setDeletionModalOpen(!deletionModalOpen);
  };

  const classes = useStyles(styles);
  if (isNilOrEmpty(scope)) return null;

  return (
    <>
      <DeleteScopeConfirmationModal
        name={prop('name', scope)}
        confirm={() => onDelete(prop('scopeId', scope))}
        open={deletionModalOpen}
        toggle={toggleDeletionModal}
      />
      <MuiChip
        label={prop('name', scope)}
        className={clsx(classes.baseStyle, {
          [classes.statusValidChip]: isEnabled && !hasMissingParameters && !isCurrentScope,
          [classes.statusValidCurrentChip]: isEnabled && !hasMissingParameters && isCurrentScope,
          [classes.statusEnableChip]: isEnabled && hasMissingParameters && !isCurrentScope,
          [classes.statusEnableCurrentChip]: isEnabled && hasMissingParameters && isCurrentScope,
          [classes.statusDisableChip]: !isEnabled && !isCurrentScope,
          [classes.statusDisableCurrentChip]: !isEnabled && isCurrentScope,
        })}
        onDelete={() => toggleDeletionModal()}
        onClick={() => onClick(scope)}
      />
    </>
  );
};

ScopeChip.propTypes = {
  engine: shape({}),
  id: number,
  handleChange: func,
};

ScopeChip.defaultProps = {
  engine: null,
  id: null,
  handleChange: Function.Prototype,
};

export default ScopeChip;
