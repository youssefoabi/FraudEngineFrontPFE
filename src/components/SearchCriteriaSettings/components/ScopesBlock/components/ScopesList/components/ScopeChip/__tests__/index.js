import { itRendersAllMutations } from '@cdiscount/jest-utils';

import ScopeChip from '../';

jest.resetModules();

jest.mock('../useScope', () => ({
  __esModule: true,
  default: jest
    .fn()
    .mockReturnValueOnce({})
    .mockReturnValueOnce({
      scope: { name: 'scope', scopeId: 1 },
      hasMissingParameters: true,
      onDelete: jest.fn(),
      onClick: jest.fn(),
      isEnabled: true,
    })
    .mockReturnValueOnce({
      scope: null,
    })
    .mockReturnValueOnce({
      scope: {},
    }),
  namedExport: 'useScope',
}));

const mutations = [
  { name: 'without props' },
  { name: 'with props' },
  { name: 'with null scope' },
  { name: 'with empty scope' },
];

describe(ScopeChip.name, () => {
  itRendersAllMutations(ScopeChip, mutations);
});
