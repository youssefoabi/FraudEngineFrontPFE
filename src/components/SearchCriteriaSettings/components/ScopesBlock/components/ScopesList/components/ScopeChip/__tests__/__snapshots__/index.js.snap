// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScopeChip ScopeChip renders correctly with empty scope 1`] = `""`;

exports[`ScopeChip ScopeChip renders correctly with null scope 1`] = `""`;

exports[`ScopeChip ScopeChip renders correctly with props 1`] = `
<Fragment>
  <DeleteEnginesConfirmationModal
    confirm={[Function]}
    name="scope"
    open={false}
    toggle={[Function]}
  />
  <WithStyles(ForwardRef(Chip))
    className="makeStyles-baseStyle-8 makeStyles-statusEnableChip-9"
    label="scope"
    onClick={[Function]}
    onDelete={[Function]}
  />
</Fragment>
`;

exports[`ScopeChip ScopeChip renders correctly without props 1`] = `""`;
